clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2015,8 +2013,38 @@ class SyclKernelIntHeaderCreator : public SyclKernelFieldHandler {
   using SyclKernelFieldHandler::handleSyclHalfType;
   using SyclKernelFieldHandler::handleSyclSamplerType;
 };
+
 } // namespace
 
+void Sema::CheckSYCLKernelCall(FunctionDecl *KernelFunc, SourceRange CallLoc,
+                               ArrayRef<const Expr *> Args) {
+  const CXXRecordDecl *KernelObj = getKernelObjectType(KernelFunc);
+  if (!KernelObj) {
+    Diag(Args[0]->getExprLoc(), diag::err_sycl_kernel_not_function_object);
```

Comment:
I think to emit such diagnostic we also should check that kernel parameter is callable object, i.e. it has operator () defined.
BTW, will it be better to emit this diagnostic at the point of `sycl_kernel` attribute handling? 
I mean I understand that this patch adds infrastructure to emit better diagnostics pointing to kernel invocation in user's code, but it is our implementation decision to require the first parameter of `sycl_kernel` function to be a function object, I don't see the value of emitting the whole instantiation stack for this particular diagnostic.

Reply:
>I think to emit such diagnostic we also should check that kernel parameter is callable object, i.e. it has operator () defined.

The body of the function actually checks that already.  When we instantiate the function, the attempt to 'call' it in the body of the sycl_kernel will ensure it is a callable.

>BTW, will it be better to emit this diagnostic at the point of sycl_kernel attribute handling?

We actually cannot do that, we don't know the callable until we try to instantiate the sycl_kernel call.  Otherwise, the 'callcable' is just the empty function template parameter.

>I don't see the value of emitting the whole instantiation stack for this particular diagnostic.

The motivation is that the call to the kernel in the library is almost always wrapped in a different call.  SO the 'location' highlighted with the error is going to be in sycl.hpp nearly every time (as @srividya-sundaram found).  For the user of the library, they have no idea where they caused it to happen (since it is wrapped in a different template!).

By forcing us to emit the template-instantiation-tree, we can make sure that at least a note shows the user where in THEIR code the probelm is.

Reply:
> The motivation is that the call to the kernel in the library is almost always wrapped in a different call. SO the 'location' highlighted with the error is going to be in sycl.hpp nearly every time (as @srividya-sundaram found). For the user of the library, they have no idea where they caused it to happen (since it is wrapped in a different template!).
By forcing us to emit the template-instantiation-tree, we can make sure that at least a note shows the user where in THEIR code the probelm is.

I get the motivation for this template-instantiation-tree emission and it is a good thing, but I'm not sure that it is the right way to diagnose parameters of the function marked with `sycl_kernel` attribute, because these functions are implementation details.

Reply:
I'm not sure what else we can do.  We don't have a good way to figure out the difference between a library call and a call from user code.  This is pretty typical of compilers with template instantiations, we error on the bad line (despite it being in a library), then give the trail to help the user figure it out.  

See here for a good example: https://godbolt.org/z/38ac41

The NICE part about this patch however, is it actually DOES produce a good way to get a better diagnosis in the future if we care to.  If we added a new attribute (say, named "calls_sycl_kernel") we could potentially do these checks in THAT case as well.  

It would allow the library to put it on their entry point to get that checking on their parameter type right away, rather than wait for the kernel invocation.

Reply:
The thing that worries me about this diagnostic is that we actually check implementation detail function.
I was going just say that we just need to use this thing for cases when we know that we want to diagnose user's code, for example when there is a problem with kernel name and so on. Then I realized that even in this case we will take some template parameter of the same implementation detail function and analyze it and this case is not different from yours...

Okay, my concern is resolved then.